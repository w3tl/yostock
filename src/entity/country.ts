import { ObjectType, Field } from "type-graphql";

@ObjectType()
export class Country {
  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  number: number;
}