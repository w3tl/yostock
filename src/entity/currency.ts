import { ObjectType, Field } from "type-graphql";

@ObjectType()
export class Currency {
  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  symbol: string;
}