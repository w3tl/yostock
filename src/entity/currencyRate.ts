import { InputType, Field, GraphQLISODateTime, ObjectType } from "type-graphql";
import { Entity, ManyToOne, PrimaryColumn, Column, RelationId, Index } from "typeorm";
import { CurrencyRateProvider } from "./currencyRateProvider";

@ObjectType()
@Entity()
@Index(["providerId", "from", "to", "validFrom"], { unique: true })
export class CurrencyRate {
  @PrimaryColumn({ type: "varchar", length: 3 })
  @Field()
  from: string;

  @PrimaryColumn({ type: "varchar", length: 3 })
  @Field()
  to: string;

  @PrimaryColumn({ type: "date" })
  @Field()
  validFrom: string;

  @Column({ type: "float" })
  @Field()
  rate: number;
  
  @Field(() => CurrencyRateProvider)
  @ManyToOne(() => CurrencyRateProvider)
  provider: CurrencyRateProvider;

  @PrimaryColumn({ type: "varchar", length: 10 })
  @RelationId((rate: CurrencyRate) => rate.provider)
  providerId: string;
}

@InputType()
export class CurrencyRateUpdateInput {
  @Field()
  providerId: string;

  @Field(() => GraphQLISODateTime)
  fromDate: Date;

  @Field(() => GraphQLISODateTime)
  toDate: Date;
}

@ObjectType()
export class CurrencyRateUpdatePayload {
  @Field()
  loaded: number;
}

@InputType()
export class CurrencyRateInput {
  @Field()
  fromCurrency: string;
  @Field()
  toCurrency: string;
  @Field(() => GraphQLISODateTime)
  fromDate: Date;
  @Field(() => GraphQLISODateTime, { nullable: true })
  toDate: Date;
  @Field()
  providerId: string;
}