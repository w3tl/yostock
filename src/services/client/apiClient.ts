export interface IResponse<T> {
  statusCode: number,
  result: T | string,
}

// export interface IRequestOptions {
  // responseProcessor?: Function,
  //Dates aren't automatically deserialized by JSON, this adds a date reviver to ensure they aren't just left as strings
  // deserializeDates?: boolean,
  // queryParameters?: ifm.IRequestQueryParams
// }

export interface IAPIClient {
  get<T>(url: string, params: Record<string, string>): Promise<IResponse<T>>;
}