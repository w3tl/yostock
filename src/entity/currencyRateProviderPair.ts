import { ObjectType, Field, ID } from "type-graphql";
import { Entity, ManyToOne, PrimaryColumn, Column, RelationId, Index } from "typeorm";
import { Currency } from "./currency";
import { CurrencyRateProvider } from "./currencyRateProvider";

@Entity()
@ObjectType()
@Index(["providerId", "fromCurrencyId", "toCurrencyId"], { unique: true })
export class CurrencyRateProviderPair {
  @PrimaryColumn({ type: "varchar", length: 3 })
  fromCurrencyId: string;

  @PrimaryColumn({ type: "varchar", length: 3 })
  toCurrencyId: string;

  @Field(type => Currency)
  fromCurrency: Currency;

  @Field(type => Currency)
  toCurrency: Currency;

  @PrimaryColumn({ type: "varchar", length: 10 })
  @RelationId((pair: CurrencyRateProviderPair) => pair.provider)
  providerId: string;

  @Field(type => CurrencyRateProvider)
  @ManyToOne(type => CurrencyRateProvider)
  provider: CurrencyRateProvider;
}