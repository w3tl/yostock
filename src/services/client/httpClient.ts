import fetch from "node-fetch";
import { IAPIClient, IResponse } from "./apiClient";

export class HTTPClient implements IAPIClient {
  async get<T>(url: string, params?: Record<string, string>): Promise<IResponse<T>> {
    const callURL: URL = new URL(url);
    if (params) {
      Object.keys(params).forEach(key => callURL.searchParams.append(key, params[key]));
    }

    const response = await fetch(callURL.toString());

    const resp: IResponse<T> = {
      statusCode: response.status,
      result: null,
    };

    resp.result = await response.text();

    return resp;
  }
}