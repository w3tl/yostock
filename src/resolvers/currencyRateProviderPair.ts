import { Resolver, FieldResolver, Root } from "type-graphql";
import { Inject, Service } from "typedi";
import { CurrencyRateProviderPair } from "../entity/currencyRateProviderPair";
import { Currency } from "../entity/currency";
import { ISOService } from "../services/iso";

@Service()
@Resolver(() => CurrencyRateProviderPair)
export class CurrencyRateProviderPairResolver {
  constructor(
    @Inject() private readonly isoService: ISOService,
  ) {}

  @FieldResolver()
  fromCurrency(@Root() pair: CurrencyRateProviderPair): Currency {    
    return this.isoService.getOneCurrency(pair.fromCurrencyId);
  }

  @FieldResolver()
  toCurrency(@Root() pair: CurrencyRateProviderPair): Currency {    
    return this.isoService.getOneCurrency(pair.toCurrencyId);
  }
}