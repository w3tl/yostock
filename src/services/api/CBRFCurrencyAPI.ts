import * as parser from "fast-xml-parser";
import { IAPIClient } from "../client/apiClient";
import { ICurrencyAPI } from "./currencyAPI";
import { CurrencyRate } from "../../entity/currencyRate";
import { CurrencyRateProviderPair } from "../../entity/currencyRateProviderPair";

type CBRFCurrenciesValueType = {
  Item: CBRFCurrenciesItemType[];
};

type CBRFCurrenciesItemAttributesType = {
  ID: string;
};

type CBRFCurrenciesItemType = {
  attributes: CBRFCurrenciesItemAttributesType;
  ISO_Char_Code: string;
};

type CBRFCurrenciesType = {
  Valuta: CBRFCurrenciesValueType;
};

type CBRFRateItemAttributesType = {
  Date: string,
  Id: string,
};

type CBRFRateItemType = {
  attributes: CBRFRateItemAttributesType,
  Nominal: number,
  Value: string,
};

type CBRFRateAttributesType = {
  ID: string,
  DateRange1: string,
  DateRange2: string,
  name: string,
};

type CBRFRateType = {
  attributes: CBRFRateAttributesType,
  Record: CBRFRateItemType | CBRFRateItemType[],
};

export class CBRFCurrencyAPI implements ICurrencyAPI {
  providerId: string;
  url: string;
  client: IAPIClient;

  private readonly currenciesPath = "XML_valFull.asp";
  private readonly ratesDailyPath = "XML_daily.asp";
  private readonly ratesDynamicPath = "XML_dynamic.asp";

  private readonly xmlOptions: Partial<parser.X2jOptions> = {
    attrNodeName: "attributes",
    textNodeName: "#text",
    attributeNamePrefix: "",
    arrayMode: false,
    ignoreAttributes: false,
    parseAttributeValue: false,
  };

  constructor(providerId: string, url: string) {
    this.providerId = providerId;
    this.url = url;
  }

  private getDateFormated(date: Date): string {
    return date.toLocaleDateString("en-GB", { year: "numeric", month: "2-digit", day: "2-digit" });
  }

  private async getCurrencies(): Promise<Record<string, string>> {
    // TODO: add caching
    const response = await this.client.get<string>(`${this.url}/${this.currenciesPath}`, {});    
    const { Valuta: value }: CBRFCurrenciesType = parser.parse(response.result, this.xmlOptions);

    return value.Item.reduce((prev, curr): Record<string, string> => {
      prev[curr.ISO_Char_Code] = curr.attributes.ID;
      return prev;
    }, {});
  }

  async getRates(fromDate: Date, toDate: Date, pairs: CurrencyRateProviderPair[]): Promise<CurrencyRate[]> {
    const currencies = await this.getCurrencies();
      
    let rates: CurrencyRate[] = [];
    for (const pair of pairs) {
      const response = await this.client.get<string>(`${this.url}/${this.ratesDynamicPath}`, {
        date_req1: this.getDateFormated(fromDate),
        date_req2: this.getDateFormated(toDate),
        "VAL_NM_RQ": currencies[pair.fromCurrencyId],
      });

      const values: CBRFRateType = parser.parse(response.result, this.xmlOptions).ValCurs;      
      const records: CBRFRateItemType[] = Array.isArray(values.Record) ? values.Record : [values.Record];

      rates = [...rates, ...records
        .filter(rate => rate.attributes.Id === currencies[pair.fromCurrencyId])
        .map(rate => {
          const currencyRate = new CurrencyRate();
          currencyRate.from = pair.fromCurrencyId;
          currencyRate.to = pair.toCurrencyId;
          currencyRate.validFrom = rate.attributes.Date.split(".").reverse().join("-");
          //new Date(record["attributes"].Date.split(".").reverse().join("/")),
          currencyRate.rate = parseFloat(rate.Value.replace(",", "."));
          currencyRate.providerId = this.providerId;
          
          return currencyRate;
        })];
    };    

    return rates;
  }
}
