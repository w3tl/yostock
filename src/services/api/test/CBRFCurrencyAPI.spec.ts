import { mocked } from 'ts-jest/utils';
import { CurrencyRateProviderPair } from "../../../entity/currencyRateProviderPair";
import { CBRFCurrencyAPI } from "../CBRFCurrencyAPI";
import { getCurrencies, getRates } from "./CBRFResponses";

import { HTTPClient } from "../../client/httpClient";

const mockedGet = jest.fn();

jest.mock("../../client/httpClient", () => ({
  HTTPClient: jest.fn().mockImplementation(() => ({
      get: mockedGet,
  })),
}));

describe("CBRFCurrencyAPI", () => {
  const mockedClient = mocked(HTTPClient, true);
  
  const currencyAPI = new CBRFCurrencyAPI("CBRF", "https://example.com");  
  currencyAPI.client = new HTTPClient();

  beforeEach(() => {
    mockedClient.mockClear();
   });

  it("getRates should retrive currency codes", async () => {
    mockedGet
      .mockReturnValueOnce({
        result: getCurrencies,
      })
      .mockReturnValue({
        result: getRates,
      });

    const pairs: CurrencyRateProviderPair[] = [];

    const pair = new CurrencyRateProviderPair();
    pair.fromCurrencyId = "USD";
    pair.toCurrencyId = "RUB";
    pair.providerId = "CBRF";

    pairs.push(pair);

    const rates = await currencyAPI.getRates(new Date(), new Date(), pairs);

    expect(mockedGet.call).toMatchSnapshot();
    expect(rates).toMatchSnapshot();
  });
});

// empty result in period today