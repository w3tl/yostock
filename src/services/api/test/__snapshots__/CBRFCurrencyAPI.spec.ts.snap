// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CBRFCurrencyAPI getRates should retrive currency codes 1`] = `[Function]`;

exports[`CBRFCurrencyAPI getRates should retrive currency codes 2`] = `
Array [
  CurrencyRate {
    "from": "USD",
    "providerId": "CBRF",
    "rate": 28.62,
    "to": "RUB",
    "validFrom": "2001-03-08",
  },
  CurrencyRate {
    "from": "USD",
    "providerId": "CBRF",
    "rate": 28.62,
    "to": "RUB",
    "validFrom": "2001-03-12",
  },
  CurrencyRate {
    "from": "USD",
    "providerId": "CBRF",
    "rate": 28.67,
    "to": "RUB",
    "validFrom": "2001-03-13",
  },
  CurrencyRate {
    "from": "USD",
    "providerId": "CBRF",
    "rate": 28.65,
    "to": "RUB",
    "validFrom": "2001-03-14",
  },
]
`;
